lib/recursive-methods.rb

```diff
@@ -1,49 +1,87 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+    return 1 if n == 0
+    return n if n == 1
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length == 1
+  return '' if s.empty?
+
+  return reverse(s[1..-1]) + s[0]
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+
+#inspired by interview cake reverse string solution
+def reverse_inplace(s, left = 0, right = s.length - 1)
+  return '' if s.empty?
+  return s if s.length == 1
+  if left < right
+    s[left], s[right] =
+    s[right], s[left]
+    reverse_inplace(s, left + 1, right - 1)
+  end
+  return s
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    return 0 if n == 0
+    return 2 + bunny(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def nested(s)
```

Comment:
You're creating a new array with each method call this is O(n^2) for both space and time.